scripts/list_apis.py

```diff
@@ -0,0 +1,195 @@
+
+
+def should_skip(obj: Any, relevant_module: ModuleType) -> bool:
+    module = obj
+
+    if not isinstance(obj, ModuleType):
+        module_name = getattr(obj, "__module__", None)
+
+        if module_name is None:
+            return False
+
+        module = importlib.import_module(module_name)
+
+    if module == relevant_module:
+        return False
+
+    module_file = None
+    try:
+        # For some objects merely accessing an attribute causes an error, so catch
+        # that here
+        module_file = getattr(module, "__file__", None)
+    except RuntimeError as e:
+        return False
+
+    if not isinstance(module_file, str):
+        return False
+
+    return f"/{relevant_module.__name__}/" not in module_file
+
+
+def get_name(path: List[str], name: str) -> str:
+    return ".".join(path + [name])
+
+
+def add_item(path: List[str], name: str, out: CrawlResult) -> None:
+    is_private = False
+    for item in path:
+        if item.startswith("_"):
+            is_private = True
+            break
+    if name.startswith("_"):
+        is_private = True
+
+    if is_private:
+        out.private.append(get_name(path, name))
+    else:
+        out.public.append(get_name(path, name))
+
+
+def get_module_attributes(module: ModuleType) -> List[str]:
+    attrs = []
+
+    try:
+        attrs += dir(module)
+    except Exception as e:
+        err_print(e)
+        pass
+
+    try:
+        attrs += getattr(module, "__all__", [])
+    except Exception as e:
+        err_print(e)
+        pass
+
+    return attrs
+
+
+def crawl_helper(
+    obj: Any,
+    name: str,
+    path: List[str],
+    seen_objects: Set[int],
+    out: CrawlResult,
+    relevant_module: ModuleType,
+) -> None:
+    if should_skip(obj, relevant_module):
+        return
+
+    if id(obj) in seen_objects:
+        return
+
+    seen_objects.add(id(obj))
+
+    if isinstance(obj, ModuleType):
+        attrs = get_module_attributes(obj)
+
+        attrs = sorted(list(set(attrs)))
+        add_item(path, name, out)
+
+        for attr in attrs:
+            try:
+                next_obj = getattr(obj, attr)
+                crawl_helper(next_obj, attr, path + [name], seen_objects, out, relevant_module)
+            except ModuleNotFoundError as e:
+                err_print(e)
+                err_print(f"ERROR: unknown module {get_name(path, name)}")
+                out.errors.append(
+                    CrawlError(
+                        reason="unknown module", path=get_name(path, name)
+                    )
+                )
+            except AttributeError as e:
+                err_print(e)
+                err_print(f"ERROR: unaccessible attribute {get_name(path, name)}")
+                out.errors.append(
+                    CrawlError(
+                        reason="unaccessible attribute", path=get_name(path, name)
+                    )
+                )
+
+    elif inspect.isclass(obj):
+        attrs = dir(obj)
```

Comment:
You don't add the class itself here?

Reply:
it should be (like with modules), good catch